### DISCUSSION TITLE
Custom protocol redesign

### DISCUSSION THREAD

**Discussion by majkinetor** [1Remote/1Remote#432] [Ideas]
Currently, we have custom protocols that can be defined per connection category. Also, we have App connection type which is almost exactly the same as protocol.

<details><summary>details</summary> 

Protocols:

![image](https://github.com/1Remote/1Remote/assets/85767/c5ad5128-4405-43c2-800e-de342ef5cb20)

App:

![image](https://github.com/1Remote/1Remote/assets/85767/ac00f02c-c4cc-4513-9aaf-d341194610ca)

</details>

I don't think this is good for two reasons:

There is no particular distinction between protocol and app:

1. Both are CLI programs driven by arguments and environment vars
1. Protocols are now artificially tied to categories, there is no need for that, it's not intuitive and is even misleading (FreeRDP in VNC category as one example)
2. There is currently no way to provide additional protocol settings, which is needed almost all the time
1. We can simplify interface by treating them the same

## The main idea

Remove protocols and make App better.

### Option 1

**Easy option:** remove protocols completely without adding anything new except small changes to App "connection":

Regarding this option:

1. `HOSTNAME`, `PORT`, `USERNAME` and `PASSWORD` (HPUP) must be added via command line
    1. Not a great user interface but passable. We could make *CMD parameter* a bit more user-friendly - multiline, wrapped, so it's easy to see the entire thing without scrolling or using *Preview*. There is no functionality lost here.
1. Availability detection is lost
    1. Unless we show HPUP fields in Apps too and pass them via environment vars as with other connections
    3. People may ignore them if they do not need them
    4. if `HOSTNAME` is set, we can run availability detection again
1. Alternatives are lost in its current form
    1. Again, unless we show HPUP fields
    1. ... but can be returned even without them via a set of alternative CMD lines

In summary, this is easy to achieve and will make app less complex. To implement:

1. Remove protocols from configuration and *Select runner* from each connection type
1. In the App "connection"
    1. Add HPUP, availability detection and Alternatives
    1. Add environment vars section (otherwise the feature is lost)
    1. Make *Cmd Parameter* multiline and wrapped (one is expected to manually map HPUP environment vars that 1RM creates the same as before)

### Option 2

**Complex option:** Predefine some applications, similar to what we have now with protocols:

1. Remove protocols UI in configuration completely
1. Add Apps tab in configuration similar to protocols but without connection categories
    1. Allow user to define array of apps
1. In App "connection"
    1. Add *Select App* with defined apps and remove *Select Runner* from other connection types
    1. Present mandatory parameters as input elements, so protocols can have HPUP as before
    1. 1RM will pass mandatory parameters as env vars, the same as before
    1. 1RM currently keeps protocol definition as JSON, we can extend it to support additional stuff

#### Definition

1. `name [string, mandatory]`
    1. Mandatory user supplied name of the tool
1. `cmd [string, mandatory]`
    1. Command line arguments
1. `path [string, optional]`
    1. Path list where tool will be searched for, in addition to the `$Env:PATH` which is always included
1. `environment [string, optional]`
    1. List of user provided environment vars, one per line in the form `ENV=VALUE`
1. `hosting [bool, default: true]`
    1. True if tool is to be hosted
1. `elevated [bool, default: false]`
    1. True if tool is to be run with admin credentials
    1. **NOTE**: If multiple elevated tools are run at the same time, it would be good to ask for elevation only once or to cache elevation response for some time.
1. `helpUrl [string, optional]`
    1. URL of the help page of the tool
1. `flags [array[object], optional]`
    1. Detailed list of the tool flags that could be offered by 1RM as a list when creating connection
    1. Properties
        1. `name [string, mandatory]` - Name of the argument (and env. var that 1RM will use to pass value)
        1. `format [string, mandatory]` - Format of the CLI expression
        1. `mandatory [bool]` - Is this CLI parameter mandatory
        1. `help [string]` - Help tip
    1. **NOTES**:
        1. One can add HPUP here and mark them as mandatory in order for 1RM to show them.
        1. We could have those mandatory parameters shown as alternatives
        1. To get availability detection, we could have another bool on the flags, `hostname: true`

We could have extra fields in definition, such as Icon URL, flag type, name on multiple languages etc... all are trivial to add any time.

The purpose of this is to be able to share tools like now (and maybe even have a list of officially pre-added tools) and to make adding new App more user friendly instead of just having one *Cmd Paramter* field. Yes, its some work, although not much but makes it more professional and its IMO much better for end users because they will either get definition from the Internet or it will come included eventually with 1RM.

Example:

```json
[
  {
    "name": "TigthVNC",
    "cmd" : "tvnviewer.exe",
    "path": "C:\\Program Files\\TightVNC; c:\\Program Files (86)\\TightVNC; c:\\Tools\\TightVNC",
    "helpUrl": "https://www.tightvnc.com/vncviewer.1.php",
    "env": "VNC_PASSWORD=%PASSWORD%",  //Assumes we have HPUP fields, if not, we can add env with hard coded password on App directly
    "hosting": true
  },
  {
    "name": "FreeRDP",
    "cmd": "wfreerdp.exe",
    "path": "C:\\Tools\\WFreeRDP",
    "hosting": true,
    "helpUrl": "https://github.com/awakecoding/FreeRDP-Manuals/blob/master/User/FreeRDP-User-Manual.markdown",
    "flags": [
        {
            "name": "Hostname",
            "format": "/v:%s",
            "mandatory": true
        },
        {
            "name": "Port",
            "format": "/port:%s",
            "mandatory": true,
            "default": "3389"    //We could have default, although in this case it doesn't have to be mandatory
        },
        {
            "name": "Username",
            "format": "/u:%s",
            "mandatory": true
        },
        {
            "name": "Password",
            "format": "/p:%s",
            "mandatory": true
        },
        {
            "name": "Audio mode",
            "format": "/audio-mode:%s",
            "help": "0 (local), 1 (remote)"
        },
        {
            "name": "Desktop Width",
            "format" "/w:%s"
        },
        {
            "name": "Desktop Height",
            "format" "/w:%s"
        },
        {
            "name": "Dynamic resolution",
            "format": "/dynamic-resolution"    // We could have type option in flag for bool|enum|number .... for even more user interface
        },
        {
            "name": "Drives",
            "format": "/drives",
            "help": "Set any value to see local drives remotelly"   //or just existence of field, even empty one counts as `true`
        },
        {
            "name": "Certificate Options",
            "format": "cert:%s",
            "help": "ignore, check ..."
        }
    ]
  }
]
```

#### New App

1. Add *Select App* with defined apps
    1. If not selected, we can use App as before, by manually specifying complete command line
1. Show Help link based on `HelpUrl` if it exists
1. For each mandatory flag create input UI element automatically
1. For non mandatory flags, create menu so one can add the flag via + button
1. Add alternatives for mandatory flags
1. Add extra environment vars (to be added in addition to those from definition)
1. **NOTE**: *Cmd params* are added in addition to to those provided in tool definition via `cmd` attribute

Mockup:
```
| App |

Name:
Tags:
Logo/color:

Script before/after:

Select app: [   FreeRdp  ]

Cmd line:   [ 
     /f  /admin 
]

Flags [+]
   Hostname:        [   ]
   Port:            [ 3389 ]
   Username:        [   ]  
   Password:        [   ] 
   
   Desktop Width:   [ 1200 ] x
   Desktop Height:  [ 800 ]  x

Alternative  [+]

Environment: [  
    
]
```

---
Links and related:
- https://github.com/1Remote/1Remote/issues/61#issuecomment-760897667
- https://github.com/1Remote/1Remote/issues/255
- https://github.com/1Remote/1Remote/issues/360
- https://github.com/1Remote/1Remote/issues/205

**Comment by majkinetor** [comment]
@VShawn, comments please :)

**Reply by VShawn** [reply]
Both options would disrupt the current intuitive logic of using protocols with runners(user set one connection, and choose which runner to use when start connection).

## I would not choose option2:

1. Option 2 is incomplete as it fails to implement the associated options in your plan. For example, after setting dynamic resolution, the option to set a fixed resolution should be hidden.
2. Based on 1, it would require building a strong system to generates UI (WFP can achieve this, but it's not as convenient as HTML. I'm unsure about the effort required since I haven't tried it before, and it would obviously break the MVVM architecture).
3. The built-in RDP client, VNC client, and S/FTP client in 1RM cannot be launched using "cmd [string, mandatory]".
4. When modifying custom protocols, all connection configurations for the modified protocol need to be updated. For example, when renaming a field.
5. Based on 4, if we use a team database to store connection info of the custom protocols, we need to synchronize the custom protocol configuration files within the team and it may conflict

That's all I think of for now. The most critical is that the workload for option 2 is too large, and 
I can't complete it quickly on my own. It would impact the release time for version 1.0. I suggest considering this option again after trying out web UI in the future.

## Option 1 seems ok

but we need to 

- consider how to guide users who wish to use a custom runner towards the APP protocol. 
- maybe provide templates for the APP protocol. 
- easily conver from other protocol to APP protocol. 
- APP protocol maybe renamed to `Custom`

I'll need to think about it some more.

**Reply by majkinetor** [reply]
> Both options would disrupt the current intuitive logic of using protocols with runners(user set one connection, and choose which runner to use when start connection).

I don't think that is particular loss of functionality. You would do so only while testing, not during normal work.

> For example, after setting dynamic resolution, the option to set a fixed resolution should be hidden.

This is not comprehensive UI, its just convenient CLI helper so you don't have to handle large command lines yourself. Its really pain in the ass (I experimented a lot with FreeRDP and its really tedious without some form of helper). 

Regarding mutually exclusive fields, as you can do the same in CLI, you can do it also on the UI. I would make this is as un-complex as possible if we are to go with this, for the start. We can always improve on it later. 

> The built-in RDP client, VNC client, and S/FTP client in 1RM cannot be launched using "cmd [string, mandatory]".

This is not for built-ins, they should work as is. This is relevant just for the App connection.

> When modifying custom protocols, all connection configurations for the modified protocol need to be updated. For example, when renaming a field.

Yes, but why would you modify it? CLI tools do not change parameters names. Renaming a field would work too - changing a *format* of the field would not. 

> Based on 4, if we use a team database to store connection info of the custom protocols, we need to synchronize the custom protocol configuration files within the team and it may conflict

Its just an app (protocol or not). Current apps also suffer from this - imagine you have a different app on team computer with the same name or newer version that is not compatible. 

I was thinking about team sharing protocols, but it doesn't seem to be worth the trouble. People can synchronize manually by taking a JSON array and importing it... Its not only that, apps need to be installed, you can't just sync definition. We could provide in app definition install script to do so, but that is huge security issue (we could work around it though, with some prompts/warning/officially-approved-scripts etc.) but that all introduces complexity that could always be developed later, when we get production insights and how much is feature actually used. 

> maybe provide templates for the APP protocol.

The Option 2 is form of a template :) And why would you need a template, you can always treat any App connection as a template as you can copy it.

> easily convert from other protocol to APP protocol.

Why? 1RM doesn't have to fulfil any promise from PRM.

**Reply by VShawn** [reply]
> This is not for built-ins, they should work as is. This is relevant just for the App connection.

but in option2 you said `Remove protocols UI in configuration completely` I think that means built-ins will be involved into option2

> And why would you need a template, you can always treat any App connection as a template as you can copy it.

Yes duplicate can be tread as template, but without a template it will be painful when creating the first connection. 

My `template` means:

![image](https://github.com/1Remote/1Remote/assets/10143738/fa0082f3-c513-483b-bb4c-5bbb72477e1e)

> Why? 1RM doesn't have to fulfil any promise from PRM.

because I may create a SFTP connect today and want to turn in into WinSCP someday, if there a conver to conver built-in SFTP into APP protocol, it would be nice. Obviously, for a new user, it's easier to first create an SFTP connection and then realize that they can use WinSCP to initiate the SFTP.

**Reply by majkinetor** [reply]
>  I think that means built-ins will be involved into option2

Nah, I thought only about custom protocols.

> because I may create a SFTP connect today and want to turn in into WinSCP someday

Even today 1RM keep existing fields when switching to another connection. This could be also done when switching to App, after which you could select WinScp "template".

That auto filled CLI you mention is exctly one of the points of Option2. In above case given the definition of Winscp "Login with password" would be specified with 4 HPUP flags instead of hardly readable sentence you shown in the screenshot. For this particular case we would need combination of flags into a single CLI format though, I could easily give example, if you are up for going with it. The above example is not meant to be exhaustive, just idea at this time. In any case, "Login with password" would be typical "calculated" field based on other fields (in this case HPUP).

**Comment by majkinetor** [comment]
So I see you did harder option.

Where is the app definition stored? I see that when I type wfreerdp.exe, CLI definition pops up.

I know its preliminary but here are some bugs:

- Hint overlapping with err <details><summary>details</summary> 
![image](https://github.com/1Remote/1Remote/assets/85767/6d19ea47-f9a3-4500-aa0b-71eb38f7ca7d)</details>
- Preview does nothing for me (let me just copy text)<details><summary>details</summary> 
![image](https://github.com/1Remote/1Remote/assets/85767/4b6df3a1-0c1d-4385-8586-0575c89f247e)</details>
- Name doesn't accept space, but it can be copy pasted <details><summary>details</summary> 
![image](https://github.com/1Remote/1Remote/assets/85767/2ae78e9c-fb00-48fa-971a-5d30be6ed65b)</details>
- Flag can not be added<details><summary>details</summary> 
![image](https://github.com/1Remote/1Remote/assets/85767/0f99194c-d0bb-4e46-a071-ee82a320fd19)
![image](https://github.com/1Remote/1Remote/assets/85767/2b72091a-4364-4d51-85b0-f25ac49fcfba)</details>
- Errors stay on top <details><summary>details</summary> 
![image](https://github.com/1Remote/1Remote/assets/85767/5483dd76-d980-4d82-a1e9-41ccae6ecfb8)</details>

Also, I think this is spam

![image](https://github.com/1Remote/1Remote/assets/85767/946f22a6-6720-4349-8ad6-8e65b0f5ce4c)

Also, what does Protocol mean ?

**Reply by VShawn** [reply]
Yeah, just the raw version, still got a lot of issues. I'm pretty busy this week, so I plan to do some light refactoring when I am free. Will continue working on the app protocol next month.

## Preview

The purpose of "Preview" is to preview and test(by the run button) whether the command line is correct.

## moving button

and moving button does make sense for some cli like winscp:

On WinSCP, you can't arbitrarily change the order of command-line parameters. Therefore, when editing the protocol, it may be necessary to adjust the order of the parameters.

as the gif shows chaning order will break the cli rule:


https://github.com/1Remote/1Remote/assets/10143738/b9087fad-cebd-4768-9926-cb5d66337a9c

## Protocol 

Protocol session allow user renamed the protocol name.

![image](https://github.com/1Remote/1Remote/assets/10143738/b76cec15-490f-4d51-926a-5a8263726cce)

**Reply by majkinetor** [reply]
I understood all the things except protocol :)

Regarding preview, it still does nothing for me, it just opens empty shell. Thats why I said that it would be good to give option to copy it as text.

**Reply by majkinetor** [reply]
So where is the app definition currently stored? 

![image](https://github.com/1Remote/1Remote/assets/85767/873e1e06-23a6-495c-8c55-521fdbe013e2)

**Reply by VShawn** [reply]
> So where is the app definition currently stored?

Currently, they are written in the assets and cannot be customized yet. I hope that in the future, users will be able to add new entries through a duplicate button, although I haven't fully figured out the details yet.

> I understood all the things except protocol :)

It's just to ensure the protocol name is displayed correctly. For example, if you have configured "FreeRDP" you would like its protocol to be displayed as "RDP" instead of "APP."
